File: week6/sentimental-credit/credit.py
# TODO
from cs50 import get_string,get_int

def main():
    num=get_string("number: ")
    # //american express

    # if num%(pow(10,13)) in [34,37] :#and check(num)=="true":
    if len(num)==15 and num[0] == '3' and (num[1] == '4' or num[1] == '7') and check (num):
        print("AMEX")


    # //master
    # elif num%(pow(10,14)) in range(51,56,1):# and check(num)=="true":
    elif (len(num) ==16) and (num[0]=='5') and (int(num[1])in range(6))and check (num):
        print("MASTERCARD")

    # //visa
    # elif 4 in [(num%(pow(10,12))),(num%(pow(10,15)))] :#and check(num)=="true":
    elif (len(num) in [13,16]) and (num[0]=='4')and check (num):
        print("VISA")
    else:
        print("INVALID")
        # print(num[0])
        # print(num[1])
        # print(len(num))
        # print(int(num)%(pow(10,11)))
        # last=2.7
        # print(last)
        # print(int(last))
        # check(num)


def check(num):
    sum=0
    for i in range  (len(num) - 2, -1 , -2):
        # print('debgging 1:', num[i])
        x=int(num[i])*2
        if x//10 > 0:
            sum+=(x%10+1)
        else:
            sum+=x
        # print('debuggin 2:', sum)
    for i in range  (len(num) - 1, -1, -2):
        # print('debugging 3:', num[i])
        sum+=int(num[i])
        # print('debugging 4:', sum)
    # print(sum)
    return sum%10 == 0


    # last = sum%10
    # if (last % (int(last))==0):
    #     # print("true")
    #     return 0
    # else:
    #     # print("false")
    #     return 1

main()

File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool is_valid(long num) {
    int sum = 0;
    bool multiplied = false;
    // 1. loop until the number is zero
    while(num > 0) {
        // 2. get the modulo 10 for the number
        int remainder = num % 10;
        // 3. accumulate number
        if (multiplied == false) {
            sum += remainder;
        } else {
            int temp = remainder * 2;
            sum += temp / 10 + temp % 10;
        }
        // 3. divide the number by 10
        num = num / 10;
        multiplied = !multiplied;
    }
    // printf("total: %d\n", sum);
    if (sum % 10 == 0) {
        return true;
    } else {
        return false;
    }
}

// function getting length
int calc_length(long num) {
    int len = 0;
    while(num > 0) {
        num /= 10;
        len += 1;
    }
    return len;
}

// get the last 2 digits
int get_last_2(long num) {
    while(num > 100) {
        num /= 10;
    }
    return num;
}

int get_last_1(long num) {
    while(num > 10) {
        num /= 10;
    }
    return num;
}

int main(void)
{
    // 1. GET NUMBER
    long num = get_long("Number: ");
    int len = calc_length(num);
    // printf("%ld\n", num);
    // printf("%d\n", len);

    // find visa 13 num
    if (get_last_1(num) == 4 && (len == 13 || len == 16)) {
        if (is_valid(num) == true)
            printf("VISA\n");
        else
            printf("INVALID\n");
    }
    // find express 15 num
    else if (len == 15 &&  (get_last_2(num)==34 || get_last_2(num)==37)) {
        printf("ana hna!!\n");
        if (is_valid(num) == true)
            printf("AMEX\n");
        else
            printf("INVALID\n");
    }
    // visa or master 16
    // else if(len == 16 &&  (get_last_2(num) > 50 && get_last_2(num) < 56)) {
    //     if (is_valid(num) == true)
    //         printf("MASTERCARD\n");
    //     else
    //         printf("INVALID\n");
    // } else {
    //     printf("INVALID\n");
    // }
}